full-stack-intro/client/src/components/Todos.tsx

```diff
@@ -0,0 +1,155 @@
+import { useEffect, useState } from 'react';
+import { PageTitle } from './PageTitle';
+import { TodoList } from './TodoList';
+import { TodoForm } from './TodoForm';
+
+export type UnsavedTodo = {
+  task: string;
+  isCompleted: boolean;
+};
+export type Todo = UnsavedTodo & {
+  todoId: number;
+};
+
+export function Todos() {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [isLoading, setIsLoading] = useState(true);
+  const [error, setError] = useState<unknown>();
+
+  /* Implement useEffect to fetch all todos. Hints are at the bottom of the file. */
+  useEffect(() => {
+    async function getToDos() {
+      try {
+        const resp = await fetch('/api/todos');
+        if (!resp.ok) throw new Error('Network call error');
+        const toDosJSON = await resp.json();
+        setTodos(toDosJSON);
+      } catch (err) {
+        setError(err);
+      } finally {
+        setIsLoading(false);
+      }
+    }
+    getToDos();
+  }, []);
+
+  /* Implement addTodo to add a new todo. Hints are at the bottom of the file. */
+  async function addTodo(newTodo: UnsavedTodo) {
+    try {
+      const options = {
+        method: 'POST',
+        body: JSON.stringify(newTodo),
+        headers: { 'Content-Type': 'application/json' },
+      };
+      const resp = await fetch('/api/todos', options);
+      if (!resp.ok) throw new Error('Network call Error');
+      const responseJSON = (await resp.json()) as Todo;
+      const updatingTodos = todos.concat([responseJSON]);
+      setTodos(updatingTodos);
+    } catch (err) {
+      setError(err);
+    }
+  }
+
+  /* Implement toggleCompleted to toggle the completed state of a todo. Hints are at the bottom of the file. */
+  async function toggleCompleted(todo: Todo) {
+    try {
+      todo.isCompleted = !todo.isCompleted;
+      const options = {
+        method: 'PUT',
+        body: JSON.stringify(todo),
+        headers: { 'Content-Type': 'application/json' },
+      };
+      const resp = await fetch(`/api/todos/${todo.todoId}`, options);
+      if (!resp.ok) throw new Error('Network call Error');
+      const responseJSON = (await resp.json()) as Todo;
+      const replacingIndex = todos.findIndex((singleTodo) => {
+        return singleTodo.todoId === todo.todoId;
+      });
+      const updatingTodos = todos.map((singleTodo, index) => {
+        if (index === replacingIndex) {
```

Comment:
You could, of course, just check todoIds here instead of taking two steps to find the index, then copy the array.